pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -598,6 +598,24 @@ func (configgen *ConfigGeneratorImpl) buildInboundFilterchains(in *plugin.InputP
 	matchingIP string, clusterName string, passthrough bool) []*filterChainOpts {
 	mtlsConfigs := getMtlsSettings(configgen, in, passthrough)
 	newOpts := []*fcOpts{}
+	isHybrid := false
+	if listenerOpts.tlsSettings != nil {
+		downstreamTLSContext := configgen.BuildListenerTLSContext(listenerOpts.tlsSettings, in.Node, istionetworking.TransportProtocolTCP)
+		gatewayTLSConfig := &plugin.MTLSSettings{
+			Port: in.ServiceInstance.Endpoint.EndpointPort,
+			Mode: model.MTLSPermissive,
+			HTTP: downstreamTLSContext,
```

Comment:
Setting `HTTP` here doesn't seem right. Supported modes are:
* `TLS`: tcp over TLS
* `HTTPS`: http over TLS

This is also a bit of an abuse since this is *m*TLS settings. The end result is also wrong, as we end up with a filter chain match like: `{"destinationPort":80,"transportProtocol":"tls","applicationProtocols":["http/1.0","http/1.1","h2c"]}` - match `h2c` over TLS doesn't seem right, should be `h2`?

Current config looks like:
```
{"fcm":{"destinationPort":80,"transportProtocol":"raw_buffer"},"tls":"default"}
{"fcm":{"destinationPort":80,"transportProtocol":"tls","applicationProtocols":["http/1.0","http/1.1","h2c"]},"tls":"file-cert:/etc/certs/user/tls.crt~/etc/certs/user/tls.key"}
{"fcm":{"destinationPort":80,"transportProtocol":"tls","applicationProtocols":["istio","istio-peer-exchange","istio-http/1.0","istio-http/1.1","istio-h2"]},"tls":"default"}
{"fcm":{"destinationPort":80,"transportProtocol":"tls"},"tls":"default"}
```
(from `istioctl pc l svc/echo --port 15006 -ojson | jq '.[].filterChains[] | {fcm: .filterChainMatch, tls: .transportSocket.typedConfig.commonTlsContext.tlsCertificateSdsSecretConfigs[0].name}' -c | sort | jq -c`)

So:
* plaintext we terminate mTLS. This cannot work.
* For tls with `applicationProtocols":["http/1.0","http/1.1","h2c"]` we terminate with user certs. 
* For anything else we use mTLS.

I am not sure how we should handle plaintext. It seems like it should be allowed, provided they are in STRICT mode. That makes the `tls` config additive. 

`tls` should not required specific ALPNs, IMO. If they are Istio ALPNs, we can treat it as mTLS. if not, it should use the user config.

Ideally all the cases of: `{incoming TLS, HTTPS, HTTP, TCP} x {ALPN} x {Sidecar config}` are clearly defined - probably in `TestInbound` of sidecar_simulation_test.go

Reply:
@howardjohn , I made the changes based on our offline conversations.
So, unless the PeerAuth for is set to DISABLE either ns wide or on the specific port, the TLS settings won’t take effect.
Something like this
```
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: httpbin-pa
  namespace: app1
spec:
  selector:
    matchLabels:
      app: httpbin
  mtls:
    mode: STRICT
  portLevelMtls:
    80:
      mode: DISABLE
```
in this ☝️  case, the filterchain would look like this
```
{"fcm":{"destinationPort":15006},"tls":null}
{"fcm":{"destinationPort":80,"transportProtocol":"tls"},"tls":"file-cert:/etc/istio/sidecargateway-certs/tls.crt~/etc/istio/sidecargateway-certs/tls.key"}
```
I dropped the plaintext match
If the PeerAuth is set to STRICT/PERMISSIVE NS wide, then TLS won't take effect. The filterchain match would look something like this.
```
{"fcm":{"destinationPort":15006},"tls":null}
{"fcm":{"destinationPort":80,"transportProtocol":"raw_buffer"},"tls":null}
{"fcm":{"destinationPort":80,"transportProtocol":"tls","applicationProtocols":["istio","istio-peer-exchange","istio-http/1.0","istio-http/1.1","istio-h2"]},"tls":"default"}
```
for STRICT you wont have ALPN match) 🙂  … basically it works as is
Please let me know if this makes sense